Framework/Crystal/src/SelectCellWithForm.cpp

```diff
@@ -97,6 +101,56 @@ Kernel::Matrix<double> SelectCellWithForm::DetermineErrors(std::vector<double> &
     return newUB1;
 }
 
+void SelectCellWithForm::ApplyTransform(Kernel::Matrix<double> &newUB, IPeaksWorkspace_sptr &ws, double tolerance,
+                                        int *num_indexed, double *average_error) {
+  // Try to optimize(LSQ) to find lattice errors
+  // UB matrix may NOT have been found by unconstrained least squares optimization
+  std::vector<double> sigabc(6);
+  int n_peaks = ws->getNumberPeaks();
+
+  std::vector<V3D> miller_indices;
+  std::vector<V3D> q_vectors;
+
+  std::unique_ptr<Geometry::OrientedLattice> o_lattice =
+      std::make_unique<OrientedLattice>(ws->mutableSample().getOrientedLattice());
+  Matrix<double> UB = o_lattice->getUB();
+
+  DblMatrix UBinv = newUB;
+  UBinv.Invert();
+
+  DblMatrix Tref = UBinv * UB;
+
+  o_lattice->setUB(newUB);
+
+  DblMatrix modHKL = o_lattice->getModHKL();
+
+  modHKL = Tref * modHKL;
+
+  o_lattice->setModHKL(modHKL);
+
+  DblMatrix modUB = o_lattice->getModUB();
+
+  for (int i = 0; i < n_peaks; i++) {
+    IPeak &peak = ws->getPeak(i);
+    q_vectors.emplace_back(peak.getQSampleFrame() - modUB * peak.getIntMNP() * 2 * M_PI);
```

Comment:
How well does this logic work for modulations that double the unit cell along any dimension (e.g. q=[0.5, 0, 0]) - in this case the deviation from the predicted position that goes into the error determination will depend on the parent peak taken? E.g.  a peak at (1.4, 0, 0) could have a deviation dH = +/- 0.1

Reply:
So (1.4,0,0)-(0.5,0,0) equals (0.9,0,0) which is (-0.1,0,0) away from (1,0,0) and (1.4,0,0)+(0.5,0,0) equals (1.9,0,0) which is (-0.1,0,0) away from (2,0,0)?

I think this is covered by the sign of (m,n,p) which should tell us the direction the parent is.

Reply:
I suppose one should index in the soubled unit cell, but in the case above there are two parents right? (100) and (200) - if the sign of mnp indicates the parent is (100) then the deviation dH = -0.1 would indicate the lattice parameter $a$ is overesitmated), whereas is mnp indicates the parent is (200) then it looks like $a$ is underestimated? Maybe it all averages out in the end?

Reply:
Ah OK sorry I have jusr re-read your comment. So it will add or subtract the nominal modulation 0.5 vector and for either parent the deviation will always be dH = -0.1 - I think this is the correct behaviour for example if a was too large.